Error handling for invalid CoverageContext

hmmr opened the pull request:
Addresses https://github.com/basho/riak_kv/issues/1411 (the issue is now closed by @lukebakken by sanitizing the parameter in question on the client side; this PR does the same on the server side, for the great good).

Now, instead of crashing with a `badarg`, bad (non-binary) and doctored (failing an internal CRC check in https://github.com/basho/riak_kv/blob/riak_ts-develop/src/riak_kv_pb_coverage.erl#L108) values are properly diagnosed and reported.

Commit:
sanitize CoverageContext parameter for query in qry_compiler:compile

Now, instead of crashing with a badarg, bad (non-binary) and doctored
values are properly diagnosed and reported.

Also, accept <<>> for Cover as substitutes for 'undefined'.

andytill commented:
Inputs should really be validated at the edge of the system, so at `riak_kv_ts_svc` and `riak_kv_qry_compiler` should have a function guard on the exported function to check the input and throw a `function_clause` if the input is bad.

Commit:
validate SELECT query Cover closer to external API entry point

hmmr commented:
@andytill Agree on the general view of where to sanitize input. Just wanted to cut some corners and do the validation in one place (there are two reportable errors: one, to simply check that it is a binary, and the other, that there is a valid term inside). Now the basic validation is done as early as we see `Cover` in riak_kv_ts_svc.

andytill commented:
I meant to write validate rather than sanitise in my original comment, updated it soon after but I don't think it shows history.

We should be validating the input rather than transforming it into something we might be able to process.

lukebakken commented on src/riak_kv_qry_compiler.erl:
```diff
@@ -827,12 +829,17 @@ rewrite2([#param_v1{name = [FieldName]} | T], Where1, Mod, Acc) ->
-    {riak_kv_pb_coverage:term_to_checksum_binary(OpaqueContext),
-     {FieldName, RangeTuple}}.
+    {ok, {undefined, undefined}};
+unwrap_cover(<<>>) ->
```
I think @andytill's suggestion is good, and we shouldn't modify input like this. It wouldn't have helped in my case, anyway, as I was passing `[]` in the python client.

lukebakken commented:
Looks good to me. @andytill should review as well.

andytill commented on src/riak_kv_qry_compiler.erl:
```diff
@@ -2216,4 +2223,15 @@ negate_an_aggregation_function_test() ->
       ).
 
+-define(TINY_DDL, "create table t (a timestamp not null, primary key ((quantum(a,1,d)), a))").
+-define(TINY_SQL, "select a from t where a>0 and a<2").
```
These can go directly into the test to avoid looking up these macros.

hmmr replied:
Indeed. The macros make little sense after removal of the second eunit test.

andytill commented on src/riak_kv_qry_compiler.erl:
```diff
@@ -2216,4 +2223,12 @@ negate_an_aggregation_function_test() ->
         finalise_aggregate(Select, [3])
       ).
 
+doctored_coverage_context_test() ->
```
Is the test asserting the error, it hard to tell from the title. `invalid_coverage_context_checksum_test` might be easier to understand without looking at the entire test.

Commit:
don't treat Cover = <<>> as a valid substitute for undefined

with a trivial macro cleanup in doctored_coverage_context_test, and some
renaming.

andytill commented:
+1 99dff3b

hmmr commented:
@borshop merge

hmmr commented:
@borshop: merge
